Add support for showing all package versions. Refs #175.

nickygerritsen opened the pull request:
In #175 there was a request to show the package versions. Since that is also a very useful feature for us, I have added this functionality.

Some notes / remarks:
* I added a package details page so one can view the versions here. We could later possibly add more here (like the readme). I made anonymous organization users have access to it.
* You can get to the details page both using the dropdown menu as well as clicking the package name.
* Does showing the clipboard button make sense? For `dev-*` versions this is what you want but for other versions you might want to add `^` or `~` in front of it. Maybe only let one copy the version itself as well as a separate button to copy the package name?
* I'm not sure if my logic in the Package and Version entity make sense?
* Perhaps we should add a `stable` boolean to version so we can show the `dev-*` versions in a separate table?
* I modified some tests to also test if we have the correct versions, but I'm not sure how/if we should test the reference as well? We could do it for the artifact case, since these ZIP's should have versions. For the local path one we can't assume anything since it uses local git info.

Some screenshots:
<img width="1346" alt="image" src="https://user-images.githubusercontent.com/550145/86511647-69f00500-bdfb-11ea-9993-eaa521291349.png">
<img width="1387" alt="image" src="https://user-images.githubusercontent.com/550145/86511650-6fe5e600-bdfb-11ea-9e4f-cc08c0042ab6.png">

nickygerritsen commented:
Not sure what I should do with the remaining phpstan errors, but those are related to one of my questions about how to handle this version logic

akondas commented on src/Query/User/PackageQuery/DbalPackageQuery.php:
```diff
@@ -103,6 +104,29 @@ public function getById(string $id): Option
+    /**
+     * @return Version[]
+     */
+    public function getVersions(string $packageId): array
```
it seems to me that it should be a paged list, because the quantity may grow, but at the moment it can stay that way (we can do it in a separate PR)

akondas commented on templates/organization/packages.html.twig:
```diff
@@ -16,7 +16,7 @@
             <tr>
                 <th>Name</th>
-                <th>Version</th>
+                <th>Latest version</th>
```
:+1:

akondas commented on src/Entity/Organization/Package.php:
```diff
@@ -246,4 +256,44 @@ private function setName(string $name): void
+    /**
+     * @param string[] $encounteredVersions
+     */
+    public function removeUnencounteredVersions(array $encounteredVersions): void
```
what about put this logic to `syncSuccess` method? Now this method will be always called before `syncSuccess`, so i think we can add this `$encounteredVersions` param there.

nickygerritsen replied:
Maybe even pass the whole packages to syncsuccess and move the adding/removing logic fully there?

akondas replied:
make sense

nickygerritsen replied:
So given the below idea about storage: either we need to pass $diskStorage to `syncSuccess` or not move the whole packages there. I think the second option is better? Since otherwise we are passing services.

akondas replied:
Passing these services is not bad. Generally, we will be doing a slight change of architecture towards a more layered (domain, application, infrastructure) to get away from this symfony style. At the moment it can stay as it is :+1:

akondas commented on templates/organization/package/details.html.twig:
```diff
@@ -0,0 +1,69 @@
+                <th>Version</th>
+                <th>Released</th>
+                <th>Reference</th>
+                <th></th>
```
Maybe we can add `size` column here?

nickygerritsen replied:
Any idea how I can get the size easily? I could add a ->size to `distStorage` that returns the size of the archive?

akondas replied:
Sounds good: `Storage -> size(Dist $dist): int`

akondas commented:
Looks great to me, thanks for your time :beers:.

nickygerritsen commented:
Also a colleague of my suggested that it might be more logical to add a copy button for the `composer require` command. Shall I do the following two things:
* Add a general copy button at the top that will `composer require <packagename>`; and
* Change the copy button for each such that it will `composer require <packagename>:<version>`

?

akondas commented:
> Also a colleague of my suggested that it might be more logical to add a copy button for the `composer require` command. Shall I do the following two things:
> 
> * Add a general copy button at the top that will `composer require <packagename>`; and
> * Change the copy button for each such that it will `composer require <packagename>:<version>`
> 
> ?

This also sounds good :+1:

nickygerritsen commented:
Fixes:
* Added pagination. Wasn't too hard so why not 🤷‍♂️ 
* Added size column
* Changed logic as requested, although adding/updating versions is still in the synchronize method, size we need the distStorage
* Updated the copy button things
* Added more tests to have at least the same coverage as before 🎉

akondas commented:
@nickygerritsen After testing, I suggest you do one more thing: maybe we can add info when no versions are available:
![Screenshot from 2020-07-07 08-11-13](https://user-images.githubusercontent.com/8239917/86724930-82ddfd80-c029-11ea-96bb-2460944b7010.png)

```
To see available versions of the package, please perform synchronization again.
````
and button `Sync now`

Because currently all packages that have not been synchronized will have this page blank. The rest works fine. Good job :beers:

akondas commented:
And what do you thing about something like this:
![Screenshot from 2020-07-07 08-19-50](https://user-images.githubusercontent.com/8239917/86726265-b2d9d080-c02a-11ea-8a5b-dd6ea9058c40.png)
Light space compression. If it's better then I can add it to your PR.

nickygerritsen commented:
I will do the first thing then you can do the second, looks much better.

nickygerritsen commented:
<img width="765" alt="image" src="https://user-images.githubusercontent.com/550145/86728111-6a231700-c02c-11ea-80cb-d1aa52b176ba.png">
Tada, like this?

akondas commented:
:facepalm: Actually, we don't really need this extra table, after all, all available versions can be fetch from the metadata file `var/repo/organization/p/vendor/package.json`. You do not need to change anything during synchronization. In general, it facilitates the whole process. Just write a parser that extracts this data from this file. Sorry, I don't know why I didn't think about it before ... :thinking:

nickygerritsen commented:
> 🤦 Actually, we don't really need this extra table, after all, all available versions can be fetch from the metadata file `var/repo/organization/p/vendor/package.json`. You do not need to change anything during synchronization. In general, it facilitates the whole process. Just write a parser that extracts this data from this file. Sorry, I don't know why I didn't think about it before ... 🤔

But will it be fast enough? Also pagination is then less useful since we need to read and `json_decode` the whole file anyway? Or do you still want it for client-side render speed?

I can try to write a parser for this data.

akondas commented:
Actually, that's right, I saw 30MB files, that would kill rendering this page. Let's stay with what you managed to do now. If this does not work, we can always change this :wink:

akondas commented:
This is my little UI refactor: https://github.com/Lets-Talk-NL/repman/pull/1/files :wink:

nickygerritsen commented:
Nice, merged and squashed

akondas commented:
Thanks @nickygerritsen
